## 1.1 The Elements of Programming

Programming languages have three mechanisms for combining simple ideas to form more complex ideas:
1. *primitive expressions*, the simplest entities the language is concerned with
2. *means of combination* by which compound elements are built from simpler ones, and
3. *means of abstraction* by which compound elements can be named and manipulated as units

Programming deals with two kinds of elements: *functions* and *data* (but later we will discover they are really not so distinct).
Informally,  
1. data: stuff we want to manipulate
2. functions: descriptions of rules for manipulating data

### 1.1.1 Expressions

We start by typing *statements* into the *interpreter* that *evaluates* those statements.  

We start with an expression statement: in javascript, an *expression* followed by a semicolon. One kind of primitive expression is a number.

In [1]:
439;

439

Expressions representing numbers may be combined with *operators* such as `+` or `*` to form a *compound expression* that represents the application of a corresponding *primitive function* to those numbers:

In [2]:
137 + 439;

576

Expressions such as these, which contain other expressions as components, are called *combinations*.

When a combination is formed by an *operator* symbol in the middle, and *operand* expressions to the left and right, we call it an *operator combination*. The value of an operator combination is obtained by applying the function specified by the operator to the arguments that are the values of the operands.

We place the operator between the operands by convention, a convention known as *infix notation*. It follows mathematical notation used in school and everday life.

As in mathematics, operator combiniations can be *nested*, that is, they can have operands that themselves are operator combinations:

In [3]:
(3 * 5) + (10 - 6);

19

Parentheses are used to group operator combinations, in order to avoid ambiguities.

Javavscript also follows the conventions that multiplication and division bind more strongly than addition and subtraction, i.e, `3 * 5 + 10 / 2;` is equivalent to `(3 * 5) + (10 / 2);`. We say that `*` and `/` have *higher precedence* than `+` and `-`.

Sequences are read from left to right. Thus

```javascript
1 - 5 / 2 * 4 + 3;
```

stands for

```javascript
(1 - ((5 / 2) * 4)) + 3;
```

We say that these operators are *left-associative*.

Lastly, note that, regardless of the complexity of the expression, the interpreter always operates in the same simple cycle:
1. read a statement typed by the user
2. evauate the statement
3. print the result

We refer to this mode of operation as a *read-evaluate-print loop*, or the initialism *REPL*.

### 1.1.2 Naming and the Environment

A critical aspect of a programming language is the means it provides for using names to refer to computational objects.
We say the name identifies a constant whose *value* is the object.

The first such means we will learn in JavaScript is a *constant*. In JavaScript, we name constants with *constant declarations*.

In [4]:
const size = 2;

Once the name `size` has been associated with thbe number 2, we can refer to the value 2 by name:

In [5]:
size;

2

In [6]:
5 * size;

10

Constant declaration is our language's simplest means of abstraction, for it allows us to use simple names to refer to the results of compound operations, like so:

In [7]:
const pi = 3.14159;

const radius = 10;

const circumference = 2 * pi * radius;

circumference;

62.8318

Means of abstraction like this allow us to build up computational objects with very complex structures. We write programs to avoid having to remember and repeat the details of building up these objects every time we want to use them. Instead, we write programs that construct computational objects of increasing complexity, by building up from simple expressions, step by step.

The fact that we can associate names with values implies that the interpreter must maintain some sort of memory that keeps track of name-object pairs. This memory is called the *environment* -- more precisely, the *program environment*, as a computation may involve a number of different environments.

### 1.1.3 Evaluating operator combinations

One goal of this chapter is to isolate issues about thinking procedurally.

As a case in point, let us consider the procedure the interpreter follows when evaluating operator combinations.

* To evaluate an operator combination, do the following:
  1. Evaluate the operand expressions of the combination.
  2. Apply the function that is denoted by the operator to the arguments that are the values of the operands.

Even this simple rule illustrates some important points about procedures.  
First, observe that the first step dicates we must first perform the evaluation process on each operand. Thus, the evaluation rule is *recursive*: it includes, as one of its steps, the need to invoke the rule itself.

Recursion allows us to express succinctly what might otherwise appear to be a rather complicated process.  
For example, the following expression requires that the evaluation rule be applied to four different combinations:

In [8]:
(2 + 4 * 6) * (3 + 12);

390

We can picture this process by representing the combination in the form of a tree, as shown in figure 1.1. Each combination is represented by a *node* with *branches* corresponding to the operator and the operands of the combination stemming from it. The *terminal* nodes, with no branches stemming from them, represent either operators or numbers.

![](./figs/fig1.1.png)

If we view evaluation in terms of the tree, we can imagine that the values of the operands percolate upward, starting from the terminal nodes and then combining at higher and higher levels. **In general, we will see that recursion is a very powerful technique for dealing with hierarchical, treelike objects.** In fact, the "percolate values upward" form of the evaluation rule is an example of a general kind of process known as *tree accumulation*.

The second point we can observe about this procedure is that our rules don't tell us how to evaluate primitve expressions -- the *base case* of our recursion. We take care of the primitive cases by stipulating that. 
* the values of numerals are the numbers that they name, and
* the values of names are the objects associated with those names in the environment.

The key point here is that **the enviornment plays a role in determining the meaning of names in expressions**. In an interpreted language such as JavaScript, especially when run in an interactive REPL, it is meaningless to talk about the value of an expression such as `x + 1;` without specifying any information about what value the environment would provide for `x`.

#### Declarations are not combinations
Notice the evaluation rule above does not handle declarations, like the constant declarations we saw above.  
We should not think of evaluating `const x = 3;` as applying an equality operator to two arguments `x` and `3`. Therefore `const x = 3;` is not a combination.
#### Keywords and rules

The word `const` is a *keyword* in JavaScript, a word with a particular meaning. Because they carry these meanings, keywords in languages are always reserved and cannot be used as names (at least, not without breaking programs).

A keyword or combination of keywords in a statement instructs the interpreter to treat the statement in a special way. **Each such *syntactic form* has its own evaluation rule.**

### 1.1.4 Compound Functions

So far we have demonstrated with JavaScript some of the elements common to many programming languages:
* primitive data and functions: numbers and arithmetic operations
* means of combining operations, e.g., nesting with parentheses
* a limited means of abstraction: constant declarations that associate names with values

Next we learn about a much more powerful abstraction technique: *function declarations* that associate a name with a compound operation, so that the operation can be referred to as a unit.

We being with the idea of squaring a number. To express this idea in plain language, we might say, "To square something, take it times itself." We express this in JavaScript as:

In [9]:
function square(x) {
    return x * x;
}

The name `square` identifies our *compound function*. The function represents the operation of multiplying something by itself. Notice that the thing to be multiplied is given a local name, `x`. This is similar to the role pronouns play in natural language.

The simplest form of a function declaration is:

`function` *name*`{`*parameters*`) { return` *expression*`; }`

where *name* is a symbol to be associated with the function in the environment, and the *parameters* are the names used within the body of the function to refer to the corresponding arguments to the function.

Here we only have one parameter, but more generall we have a comma-separated list of a parameters, and in an application of the function we will likewise have a comma-separated list of arguments, the values that the parameters take on when the function is applied.

In its simplest form, the *body* of a function declaration is a single *return statement*, the keyword `return` followed by the *return expression* that will yield the value of the function application.

Having declared `square`, we can now use it in a *function application* expression.

In [10]:
square(21);

441

Function applications are--after operator combinations--the second kind of combination of expressions into larger expressions that we encounter. The general form is:

*function-expression(argument-expressions)*

where the *function-expression* of the application specifies the function to be applied to the comma-separated *argument-expressions*.

To evaluate a function application, the interpreter follows a procedure quite similar to the procedure for operator combinations described in section 1.1.3.

* To evaluate a function application, do the following:
  1. Evaluate the subexpressions of the application, namely the function expression and the argument expressions.
  2. Apply the function that is the value of the function expression to the values of the argument expressions.

To see the generality of this procedure, take for example a case where the argument expression is itself a compound expression:

In [11]:
square(2 + 5);

49

We can see that the interpreter indeed follows our procedure, evaluating the argument expression first before applying the value of the function expression to the value of the argument expression.

Observe that function applicaton expressions can also serve as argument expressions:

In [12]:
square(square(3));

81

We can also use `square` as a building block in defining other functions. For example, $x^2 + y^2$ can be expressed as:

`square(x) + square(y)`

Using this expression, we can declare a functio `sum_of_squares` that, given any two numbers as arguments, produces the sum of their squares:

In [13]:
function sum_of_squares(x, y) {
    return square(x) + square(y);
}

In [14]:
sum_of_squares(3, 4);

25

And by extension we can use `sum_of_squares` as a building block in constructing yet more functions:

In [15]:
function f(a) {
    return sum_of_squares(a + 1, a * 2);
}

In [16]:
f(5);

136

In addition to compound functions, any JavaScript environment provides primitive functions that are built into the interpreter or loaded from libraries. These additional primitive functions are used in the exact same way as compuond functions. This means that it doesn't matter when we write a compound function like `sum_of_squares` whether `square` was built into the interpreter, loaded from a library, or itself defined as a compound function.

### 1.1.5 The Substitution Model for Function Application

To evaluate a function application, the interpreter follows the procedure described in section 1.1.4. For primitive functions, this is handled by the interpreter or libraries. For compound functions, the application process is as follows:
* To apply a compound function to arguments, evaluate the return expression of the function with each parameter replaced by the corresponding argument.

We can apply this process to the functions we declared above.  
Let's evaluate the application

```javascript
f(5);
```

where `f` is the function we just declared.  
We begin by retrieving the return expression of `f`:
```javascript
sum_of_squares(a + 1, a * 2);
```
Then we replace the parameter `a` with the argument 5:
```javascript
sum_of_squares(5 + 1, 5 * 2)
```
Thus the problem reduces to the evaluation of an application with two arguments and a function expression `sum_of_squares`. Evaluating this applications involves three subproblems. We continue substituting to reduce these subproblems until we end up with two primitive expressions and a primitive function, an operator applied to them: `36 + 100`. This gives us, finally, `136`.

The process we have just described is called the *substitution model* for function application.

Two points that should be stressed:
1. The purpose of the substitution is to help us to think about function application. It is *not* a description of how the interpreter really works: typically interpreters do not evaluate function applications by manipulating the text of the function to substitute values for parameters. In practice, the "substitution" is accomplished by using a local environment for the parameters.
2. Over the course of this book, we present a sequence of increasingly elaborate models of how interpreters work, culminating with a complete implementation of an interpreter and compiler. The substitution model is only the first of these. When we address the use of functions with mutable data, we will see that the substitution model breaks down and must be replaced by a more complicated model of function evaluation.

#### Applicative order versus normal order

The procedure given in 1.1.4 is not the only way to perform evaluation. An alternative model would not evaluate arguments until their values were needed. Instaed, it would first substitute argument epxressions for parameters until it obtained an expression involving only operators and primitive functions, and would then perform the evaluation.

If we used this method, the evaluation of `f(5)` would proceed according to the sequence of expansions
```javascript
sum_of_squares(5 + 1, 5 * 2)

square(5 + 1) + square(5 * 2)

(5 + 1) * (5 + 1) + (5 * 2) * (5 * 2)
```
followed by the reductions
```javascript
6 * 6 + 10 * 10
36 + 100
136
```
This gives the same answer as our previous evaluation model, but the process is different. In particular, the evaluations of `5 + 1` and `5 * 2` are each performed twice here, corresponding to the reduction of the expression
```javascript
x * x
```
with `x` replaced, respectively, by `5 + 1` and `5 * 2`.

With the order we followed above, we would isntead have done
```javascript
sum_of_squares(5 + 1, 5 * 2)

sum_of_squares(6, 10)

square(6) + square(10)
```
and then the reduction would have proceeded as just shown.

This alternative method of fully expanding and then reducing is known as *normal-order evaluation*. The interpreter actually evaluates the arguments and then applies functions, and this is called *applicative-order evaluation*. It can be shown that, for function applications that can be modeled using substitution and that yield legitimate values, normal-order and applicative-order evaluation produce the same value. JavaScript uses applicative-order evaluation. This is partly because of the additional efficiency obtained from avoiding multiple evauations of expressions, as just demonstrated. More significantly, normal-order evaluation becomes much more complicated to deal with when we leave the realm of functions that can be modeled by substitution. On the other hand, normal-order evaluation can be an extremely valuable tool.

### 1.1.6 Conditional Expressions and Predicates

The expressive power of the class of functions that we can define at this point is very limited, because we have no way to make tests and to perform different operations depending on the result of a test.

$$
|x| = 
\begin{cases}
x \text{ if } x \ge 0 \\
-x \text{ otherwise }
\end{cases}
$$

This construct is a *case analysis* and can be written in JavaScript using a *conditional expression* as

In [17]:
function abs(x) {
    return x >= 0 ? x : -x;
}

This can be expressed in English as "If $x$ is greater than or equal to zero, return $x$; otherwise return $-x$".

The general form of a conditional expression is 

*predicate* `?` *consequent-expression* `:` *alternative-expression*

Conditional expressions begin with a *predicate*, an expression whose value evaluates to either *true* or *false*, the two distinguished *boolean* values in JavaaScript. The primitive boolean expressions are `true` and `false`. The predicate is followed by a question mark, then the *consequent-epxression*, a colon, and finally the *alternative-expression*.

To evaluate a conditional expression, the interpreter starts by evaluating the *predicate*. If the *predicate* evaluates to true, the interpreter evaluates the *consequent-expression* and returns its value as the value of the conditional. If the predicate evaluates as false, the interpreter evaluates the *alternative-expression* and returns that instead.

The word *predicate* is used for operators and functions that return true or false, as well as for expressions that evaluate to true or false. Our absolute-value function `abs` makes use of the primitive predicate `>=`.

We could specify our absolute value function so that it has more than two cases:
$$
|x| = 
\begin{cases}
&x &\text{if } x \gt 0 \\
&0 &\text{if } x = 0 \\
&-x &\text{otherwise}
\end{cases}
$$

In JavaScript, we express a case analysis with multiple cases by nesting conditional expressions as alternative experssions inside other conditional expressions:

In [18]:
function abs(x) {
    return x > 0
           ? x
           : x === 0
           ? 0
           : -x;
}

Parentheses are not needed around the alternatve expression `x === 0 ? 0 : -x` because the conditional-expression syntactic form is right-associative.

The general form of a case analysis is  
$p_1$
`?` $e_1$  
`:` $p_2$  
`?` $e_2$  
$\vdots$  
`:` $p_n$  
`?` $e_n$  
`:` *final-alternative-expression*

We call a predicate $p_i$ together with its consequent expression $e_i$ a *clause*. A case analysis can be seen as a sequence of clauses, followed by a final alternative expression. We can evaluate a case analysis using what we know about how the interpreter evaluates conditonal expressions: first the interpreter evaluates the predicate $p_1$. If its value is false, then $p_2$ is evaluated. If $p_2$'s value is also found to be false, then $p_3$ is evaluated. this process continues until a predicate $p_i$ is found whose value is true, in which case the interpreter returns the value of the corresponding consequent expression $e_i$ as the value of the case analysis. If none of the $p$s is found to be true, the value of the case analysis is the value of the final alternative expression.

In addition to primitve predicates (`>=`, `>`, `<`, `<=`, `===`, and `!==`) that are applied to numbers, there are logical composition operations that enable us to construct *compound predicates*. The three most frequently used are these:

1. $expression_1$ `&&` $expression_2$  
   This operation expresses *logical conjunction*, roughly the same as the English word "and". We assume this form to be syntactic sugar for $expression_1$ `?` $expression_2$ `: false`. Note in these definitions we assume that evaluating $expression_1$ and $expression_2$ can only produce boolean values.

2. $expression_1$ `||` $expression_2$  
   This operation expresses *logical disjunction*, roughly the same as the English word "or". We assume this form to be syntactic sugar for $expression_1$ `?` `true :` $expression_2$

3. `!` $expression$  
   This operation expresses *logical negation*, roughly the same as the English word "not". The value of the expression is true when $expression$ evaluates to false, and false when $expresssion$ evaluates to true.

Notice that `&&` and `||` are syntactic forms, not operators; their right-hand expression is not always evaluated. This is a case of what is sometimes called *syntactic sugar*.

The operator `!`, on the other hand, follows the evaluation rule of section 1.1.3. It is a *unary* operator, meaning it takes only one argument, whereas the arithemetic operators and primitive predicates discussed so far are *binary*, taking two arguments. The operator `!` precedes its argument, so we call it a *prefix operator* (as opposed to the infix operators we have seen so far). Another prefix operator is the numeric negation operator, an example of which we saw as `-x` in the `abs` function above.

Let's look at an example of how these predicates are used.  
The condition that a number $x$ be in the range $5 < x < 10$ may be expressed as

```javascript
x > 5 && x < 10
```

The syntactic form `&&` has lower precedence than the comparison operators `>` and `<`, and the conditional-expression syntactic form $\dots$ `?` $\dots$ `:` $\dots$ has lower precedence than any other operator we have encountered so far, a property we used in the `abs` functions above.

As another example, we can declare a predicate to test whether one number is greater than or equal to another as

```javascript
function greater_or_equal(x, y) {
    return x > y || x === y;
}
```

or alternatively as

```javascript
function greater_or_equal(x, y) {
    return ! (x < y);
}
```

Unary operators have higher precedence than binary operators, making the parentheses necessary in the second version of the function.

These functions, when applied to two numbers, behaves the same as the operator `>=`.

#### Exercise 1.3

In [19]:
function sum_of_squares_largest(a, b, c) {
    return a > c && b > c
           ? sum_of_squares(a, b)
           : a > b && c > b
           ? sum_of_squares(a, c)
           : sum_of_squares(b, c)
}

### 1.1.7 Example: Square Roots by Newton's Method

As introduced above, functions are much like ordinary mathematical functions. They specify values that are determined by one or more parameters. But there is an important difference between mathematical functions and computer functions, that we now illustrate.

Consider the problem of computing square roots. We can define the square-root function as  

$$
\sqrt{x} = \text{ the } y \text{ such that } y \ge 0 \text{ and }y^2 = x
$$

This describes a legitimate mathetmatical function we could use to recognize whether one number is the square root of another, or to derive facts about square roots in general. But this definition does not describe a computer function. It tells us nothing about *how* to find the square root of a given number. We can make this obvious by rephrasing the definition in pseudo-JavaScript:  

```javascript
function sqrt(x) {
    return the y with y >=0 && square(y) === x;
}
```

The difference between these two types of functions is the difference between describing properties of things and describing how to do things. We can say the first is declarative knowledge and the second is imperative knowledge. **In mathematics, we are usually concerned with declarative "what is" descriptions, whereas in computer science we are usually concerned with imperative "how to" descriptions.**

How does one compute square roots? The most common way is to use Newton's method of successive approximations. This method says that whenever we have a guess $y$ for the value of the square root of a number $x$, we can perform a simple manipulation to get a better guess, closer to the actual square root, by averaging $y$ with $x/y$.

Now let's formalize this process in terms of (computer) functions. We start with a value for the radicand--the number whose square root we are trying to compute--and a value for the guess. If the guess is good enough for our purposes, we are done; if not we must repeat the process with an improved guess. We write this as a function.

In [20]:
function sqrt_iter(guess, x) {
    return is_good_enough(guess, x)
           ? guess
           : sqrt_iter(improve(guess, x), x);
}

A guess is improved by averaging it with the quotient of the radicand and the old guess:

In [21]:
function improve(guess, x) {
    return average(guess, x / guess);
}

where

In [22]:
function average(x, y) {
    return (x + y) / 2;
}

We also have to say what we mean by "good enough." The following will do for illustration. The idea is to improve the answer until it is close enough so that its square differs from the radicand by less than a predetermined tolerance, here 0.001:

In [23]:
function is_good_enough(guess, x) {
    return abs(square(guess) - x) < 0.001;
}

Finally, we need a way to get started. For instance, we can always guess that the square root of any number is 1:

In [24]:
function sqrt(x) {
    return sqrt_iter(1, x);
}

Having typed these declarations into the interpreter, we can use `sqrt` just as we can use any function:

In [25]:
sqrt(9);

3.00009155413138

The `sqrt` program also illustrates that the simple functional language we have introduced so far is sufficient for writing any purely numerical program that one could write in, say, C or Pascal. this might seem surprising, since we have not included in our language any iterative (looping) constructs that direct the computer to do something over and over again. However, the function `sqrt_iter` demonstrates that iteration can be accomplished using no special construct other than the ordinary ability to call a function.

#### Exercise 1.6
Alyssa P. Hacker doesn’t like the syntax of conditional expressions, involving the characters `?` and `:`. “Why can’t I just declare an ordinary conditional function whose application works just like conditional expressions?” she asks.22 Alyssa’s friend Eva Lu Ator claims this can indeed be done, and she declares a conditional function as follows:

In [26]:
function conditional(predicate, then_clause, else_clause) {
    return predicate ? then_clause : else_clause;
}

Eva demonstrates the program for Alyssa:

In [27]:
conditional(2 === 3, 0, 5);

5

Delighted, Alyssa uses conditional to rewrite the square-root program:

In [28]:
function sqrt_iter_conditional(guess, x) {
    return conditional(is_good_enough(guess, x),
                       guess,
                       sqrt_iter_conditional(improve(guess, x),
                                             x));
}

In [29]:
function sqrt2(x) {
    return sqrt_iter_conditional(1, x);
}

What happens when Alyssa attempts to use this to compute square roots? Explain.

In [30]:
sqrt2(4);

RangeError: Maximum call stack size exceeded

When Alyssa rewrite `sqrt` as `sqrt_iter_conditional` using `conditional`, she changes from using the conditional expression to using a function application. According to the procedure we described in section 1.1.4 for evaluating a function, the interpreter first evaluates all elements of the applcation, including the arguments, and then applies the function to the arguments. This means that every time `sqrt_iter_conditional` calls `conditional`, it will evaluate all arguments, including another call to `sqrt_iter_conditional`. And this causes the infinite recursion. We don't have this problem with the conditional expression because the *else-clause* is only evaluated if the *predicate* evaluates to `false`.

#### Exercise 1.7
The `is_good_enough` test used in computing square roots will not be very effective for finding the square roots of very small numbers. Also, in real computers, arithmetic operations are almost always performed with limited precision. This makes our test inadequate for very large numbers. Explain these statements, with examples showing how the test fails for small and large numbers.

For very small numbers, we get similar answers for different inputs `x`.

In [31]:
sqrt(0.0000000000008);

0.031250000008525

In [32]:
sqrt(0.00000000000016);

0.031250000001705

This is because the difference between `x` and the square of our guess at $\sqrt{x}$ can be less than our hard-coded value of `0.001` and still be far from the true value of the square root.

For very large numbers, our hard-coded value of `0.001` is too small to detect when we are close to the true value of the square root, so we end up with a recursion larger than the maximum call stack size.

In [33]:
sqrt(10**10**2);

RangeError: Maximum call stack size exceeded

See better phrasing of my answers here: https://mk12.github.io/sicp/exercise/1/1.html#ex1.7
> The `is_good_enough` predicate does not work well for small numbers because the tolerance is fixed. If the number is smaller than the tolerance, the result will be hopelessly inaccurate, like measuring an atom with a yard stick. It is also inadequate for very large numbers. With limited precision, it is impossible to represent small differences between large numbers. This means the algorithm might never terminate, because the guess never satisfies `is_good_enough` no matter how many times `improve` is called.

An alternative strategy for implementing `is_good_enough` is to watch how guess changes from one iteration to the next and to stop when the change is a very small fraction of the guess. Design a square-root function that uses this kind of end test. Does this work better for small and large numbers?

In [34]:
function is_good_enough2(guess, last_guess) {
    return abs(guess - last_guess) / guess < 0.001;
}

In [35]:
function sqrt_iter2(guess, x) {
    return is_good_enough2(improve(guess, x), guess)
           ? guess
           : sqrt_iter2(improve(guess, x), x);
}

In [36]:
function sqrt3(x) {
    return sqrt_iter2(1, x);
}

For small numbers, we now get a much more accurate guess. So, yes it works better overall.

In [37]:
sqrt3(0.0000000000008);

8.944277349423199e-7

In [38]:
sqrt3(0.00000000000016);

4.00001186011135e-7

For large numbers, it works in the sense that it terminates. But the estimate is not as precise compared to small numbers because a "small" change in the difference between guesses, relative to the guess itself, can still be quite large for these large numbers.

In [39]:
sqrt3(10 ** 10 ** 2);

1.000000633105179e+50

#### Exercise 1.8
Newton’s method for cube roots is based on the fact that if $y$ is an approximation to the cube root of $x$, then a better approximation is given by the value

$$
\frac{x/y^2 + 2y}{3}
$$

Use this formula to implement a cube-root function analogous to the square-root function.

In [40]:
function improve_cubrt(guess, x) {
    return ((x / guess ** 2) + 2 * guess) / 3
}

In [41]:
function cubrt_iter(guess, x) {
    return is_good_enough2(improve_cubrt(guess, x), guess)
           ? guess
           : cubrt_iter(improve_cubrt(guess, x), x);
}

In [42]:
function cubrt(x) {
    return cubrt_iter(1, x);
}

In [43]:
cubrt(27);

3.001274406506175

### 1.1.8 Functions as Black-Box Abstractions

The function `sqrt` is our first example of a process defined by a set of mutually defined functions. Note that `sqrt_iter` is *recursive*: the function is defined in terms of itsefl. We address how this works more carefully in section 1.2. First let us consider some other points illustrated by the `sqrt` example.

Observe that the problem of computing square roots breaks up naturally into a number of subproblems. Each of these is accomplished by a separate function. The entire `sqrt` program can be viewed as a tree of functions, shown in figure 1.2, that mirrors the decomposition of the problem into subproblemns.

![fig 1.2](./figs/fig1.2.png)

What is important in this decomposition strategy is that each function accomplishes an identifiable task that can be used as a module in defining other functions. For example, when we define the `is_good_enough` function in terms of `square`, we regard the `square` function as a black box. As we define `is_good_enough`, we are not concerned with *how* the `square` function computes its result, only with the fact *that* it computes the square. We can ignore the details of *how* for now. As far as the `is_good_enough` function is concerned, `square` is a *functional abstraction*. At this level of abstraction, *any* function that computes the square is equally good. Thus the following two functions `square` are indistinguishable.

```javascript
function square(x) {
    return x * x;
}

function double(x) {
    return x + x;
}

function square(x) {
    return math_exp(double(math_log(x)));
}
```

So: a function should be able to suppress detail. Users of a function may not have written it themselves, but may have obtained it from another programmer as a black box. A user should not need to know how the function is implemented in order to use it.

#### Local names

One detail of a function's implementation that should not affect users of the function is the names that the implementer of the function chooses for the parameters. Thus, from the point of the view of a user's program, the following functions should not be distinguishable:

```javascript
function square(x) {
    return x * x;
}

function square(y) {
    return y * y;
}
```

The principle that the result of a function should not depend on parameter names may seem self-evident, but its consequences are profound. The simplest consequence is that parameter names of a function must be local to the body of the function.

Notice that the function `is_good_enough` has two parameters, `guess` and `x`.

```javascript
function is_good_enough(guess, x) {
    return abs(square(guess) - x) < 0.0001;
}
```

The function `square`, that we use to write `is_good_enough` also has an argument `x`.

```javascript
function square(x) {
    return x * x;
}
```

We see that the `x` in `is_good_enough` *must* be a different `x` than the one in `square`. Running the function `square` must not affect the value of `x` that is used by `is_good_enough`, because that value of `x` may be needed by `is_good_enough` after `square` is done computing.  
If parameters were not local to the bodies of their respective functions, then the parameter `x` in square could be confused with the parameter `x` in `is_good_enough`, and the behavior of `is_good_enough` would depend on which version of `square` we used. If this were true, `square` would not be the black box we desire.

In order to for functions to work as black boxes, the names of parameters must be *bound*. We say that a function declaration *binds* its parameters. This means that a function declaration is unchanged if a bound name is consistently renamed throughout the declaration, i.e. if you declare a function with parameter `x` and you change that name to `y` along with every occurrence of `x` in the body of the function. If a name is not bound, we say it is *free*. The set of statements for which a binding declares a name is called the *scope* of that name. In a function declaration, the bound names declared as the parameters of the function have the body of the function as their scope.

In the declaration of `is_good_enough` above, `guess` and `x` are bound names, but `abs` and `square` are free. The meaning of `is_good_enough` should be independent of the names we choose for `guess` and `x` so long as they are distinct and different from `abs` and `square`.

#### Internal declarations and block structure

We have one kind of name isolation available to us so far: the parameters of a function are local to the body of the function.

The square root program illustrates another way in which we would like to control the use of names. The existing program consistes of separate functions:

```javascript
function sqrt(x) {
    return sqrt_iter(1, x); 
} 

function sqrt_iter(guess, x) {
    return is_good_enough(guess, x) 
            ? guess 
            : sqrt_iter(improve(guess, x), x); 
}

function is_good_enough(guess, x) {
    return abs(square(guess) - x) < 0.001; 
}

function improve(guess, x) {
    return average(guess, x / guess); 
}
```

The problem with this program is that the only function that is important to users of `sqrt` is `sqrt`. The other functions only clutter up their minds. They may not declare any other function called `is_good_enough` as part of another program, because `sqrt` needs that name. This problem is especially severe in the construction of large systems by many separate programmers. For example, in the construction of a large library of numerical functions, many are computed as successive approximations and thus might have functions named `is_good_enough` and `improve` as auxiliary functions. We would like to localize the subfunctions, hiding them inside `sqrt` so that `sqrt` could coexist with other successive approximations, each having its own private `is_good_enough` function.

To make this possible, we allow a function to have internal declarations that are local that function. For example, we can write `sqrt` as:

```javascript
function sqrt(x) {
    function is_good_enough(guess, x) {
        return abs(square(guess) - x) < 0.001; 
    }
    
    function improve(guess, x) {
        return average(guess, x / guess); 
    }
        return sqrt_iter(1, x); 
    } 
    function sqrt_iter(guess, x) {
        return is_good_enough(guess, x) 
                ? guess 
                : sqrt_iter(improve(guess, x), x); 
    }
    return sqrt_iter(1, x);
}
```

Any matching pair of braces designates a *block*, and declarations inside a block are local to that block. Such nesting of declarations, called *block structure*, is basically the right solution to the simplest name-packaging problem. But there is a better idea lurking here.

In addition to internalizing the declarations of auxiliary functions, we can simply them. Since `x` is bound in the declaration of `sqrt`, the functions declared internally to `sqrt` are in the scope of `x`. So it is not necessary to pass `x` explicitly to each of these functions. Instead, we allow `x` to be a free name in the internal declarations, as shown below. Then `x` gets its value from the argument with which the enclosing function `sqrt` is called. This discipline is called *lexical scoping*.

```javascript
function sqrt(x) {
    function is_good_enough(guess) {
        return abs(square(guess) - x) < 0.001; 
    }
    
    function improve(guess) {
        return average(guess, x / guess); 
    }
        return sqrt_iter(1, x); 
    } 
    function sqrt_iter(guess) {
        return is_good_enough(guess, x) 
                ? guess 
                : sqrt_iter(improve(guess, x), x); 
    }
    return sqrt_iter(1);
}
```

We will use block structure extensively to help us break up large programs into tractable pieces. The idea of block structure originated with the programming language Algol 60. It appears in most advanced programming languages and is an important tool for helping to organize the construction of large programs.

In [ ]:
## 1.2 Functions and the processes they generate

